packages/next/build/analysis/get-page-static-info.ts

```diff
@@ -0,0 +1,119 @@
+import type { PageRuntime } from '../../server/config-shared'
+import type { NextConfig } from '../../server/config-shared'
+import { tryToExtractExportedConstValue } from './extract-const-value'
+import { parseModule } from './parse-module'
+import { promises as fs } from 'fs'
+
+export interface PageStaticInfo {
+  runtime?: PageRuntime
+  ssg?: boolean
+  ssr?: boolean
+}
+
+/**
+ * For a given pageFilePath and nextConfig, if the config supports it, this
+ * function will read the file and return the runtime that should be used.
+ * It will look into the file content only if the page *requires* a runtime
+ * to be specified, that is, when gSSP or gSP is used.
+ * Related discussion: https://github.com/vercel/next.js/discussions/34179
+ */
+export async function getPageStaticInfo(params: {
+  nextConfig: Partial<NextConfig>
+  pageFilePath: string
+  isDev?: boolean
+}): Promise<PageStaticInfo> {
+  const { isDev, pageFilePath, nextConfig } = params
+
+  const fileContent = (await tryToReadFile(pageFilePath, !isDev)) || ''
+  if (/runtime|getStaticProps|getServerSideProps/.test(fileContent)) {
+    const swcAST = await parseModule(pageFilePath, fileContent)
+    const { ssg, ssr } = checkExports(swcAST)
+    const config = tryToExtractExportedConstValue(swcAST, 'config') || {}
+    if (config?.runtime === 'edge') {
+      return {
+        runtime: config.runtime,
+        ssr: ssr,
+        ssg: ssg,
+      }
+    }
+
+    // For Node.js runtime, we do static optimization.
+    if (config?.runtime === 'nodejs') {
+      return {
+        runtime: ssr || ssg ? config.runtime : undefined,
+        ssr: ssr,
+        ssg: ssg,
+      }
+    }
+
+    // When the runtime is required because there is ssr or ssg we fallback
+    if (ssr || ssg) {
+      return {
+        runtime: nextConfig.experimental?.runtime,
+        ssr: ssr,
+        ssg: ssg,
+      }
```

Comment:
**question:** if the config runtime is `edge`, but there is no SSR nor SSG, shouldn't we return `{ssr: false, ssg: false}`?

If yes, then the whole part could be shrink into
```js
return {
  runtime: ssr || ssg ? config?.runtime || nextConfig.experimental?.runtime,
  ssr, 
  ssg 
}
```